### PR TITLE
Merger the eigen implementation with master. 

### PR DESCRIPTION
**Pull request by ChiZhangatTUM**
The main modifications is to move the basic vector data(vector, matrix) from Simbody to Eigen.

### CODE REVIEW

**Comment by BenceVirtonomy** [comment]
@ChiZhangatTUM Could you pull the master first before we review it, please? Right now the diff is based on an old version.

**Comment by ChiZhangatTUM** [comment]
> @ChiZhangatTUM Could you pull the master first before we review it, please? Right now the diff is based on an old version. @Xiangyu-Hu



> @ChiZhangatTUM Could you pull the master first before we review it, please? Right now the diff is based on an old version. @Xiangyu-Hu

Now is Okay?

**Comment by BenceVirtonomy** [comment]
I don't see a new commit. You need to pull the master into your branch/ rebase your branch to the latest master + resolve all the conflicts that are shown here.

**Comment by ChiZhangatTUM** [comment]
> I don't see a new commit. You need to pull the master into your branch/ rebase your branch to the latest master + resolve all the conflicts that are shown here.

Now the conflicts are solved by merging master into the eigen implementation.

**Comment by ChiZhangatTUM** [comment]
I have change a lot bit of the CMakelists.txt, please have a double check.

File: CMakeLists.txt [hunk 1]
```diff
@@ -31,8 +31,8 @@ target_include_directories(sphinxsys_core INTERFACE ${Simbody_INCLUDE_DIR})
 target_link_directories(sphinxsys_core INTERFACE ${Simbody_LIB_DIR})
 target_link_libraries(sphinxsys_core INTERFACE ${Simbody_LIBRARIES})
 
-find_package(Eigen3 3.3 REQUIRED NO_MODULE)
 if(TARGET Eigen3::Eigen)
+    find_package(Eigen3 3.3 REQUIRED NO_MODULE)
```
Comment (FabienPean-Virtonomy) [inline CMakeLists.txt:35]: This effectively prevents to find Eigen installed elsewhere, ok considering the broken current state. The repository was not compiling before that, which means the version bundled with the CI of Eigen is not sufficient. Features used here means it requires at least Eigen 3.4

File: SPHINXsys/src/for_2D_build/adaptations/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${PROJECT_SOURCE_DIR}/cmake) # main (top) cmake dir
+include(Headersearch)
+
+file(GLOB BASE_DIR_HEADERS *.h)
+DIR_INC_HEADER_NAMES("${BASE_DIR_HEADERS}" BASE_DIR_HEADER_NAMES)
+#message("${BASE_DIR_HEADER_NAMES}")
+INSTALL(FILES ${BASE_DIR_HEADER_NAMES} DESTINATION 2d_code/include)
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/for_2D_build/adaptations/CMakeLists.txt:7]: Why new files here? Those are not related to the Eigen switch

File: SPHINXsys/src/for_2D_build/adaptations/adaptation_supplementary.cpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+			{
+				Vec2d particle_location(Real(i) * particle_spacing, Real(j) * particle_spacing);
+				Real distance = particle_location.norm();
+				if (distance < cutoff_radius)
+					sigma += kernel_ptr_->W(distance, particle_location);
+			}
+		return sigma;
+	}
+    //=================================================================================================//
+}
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/for_2D_build/adaptations/adaptation_supplementary.cpp:23]: Why new files here? Those are not related to the Eigen switch

**Review by FabienPean-Virtonomy — COMMENTED** [review]
This PR contains several significant changes beyond what it is advertised for: new files, new examples, formatting, renaming of existing classes. Please restrict the PR to the changes needed to switch to Eigen (preferred practice). Or at least report an extensive changelog of what is in here.

**Comment by BenceVirtonomy** [comment]
@ChiZhangatTUM it seems to me you reverted some previous changes when you resolved the conflicts.

**Comment by ChiZhangatTUM** [comment]
> @ChiZhangatTUM it seems to me you reverted some previous changes when you resolved the conflicts. I'm with Fabien here, I would start a new branch from the master and cherry-pick only the Eigen-related changes.

Hi, The man changes are related to eigen implementation and add new DFG funding NO, If i am wrong, please tell me.

File: SPHINXsys/src/for_2D_build/particle_dynamics/solid_dynamics/contact_dynamics_supplementary.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+						sigma += corrected_W_ij * particle_spacing_ * 0.5 * three_gaussian_weights_[i];
+					}
+				}
+				contact_density_i += 0.005 * sigma * calibration_factor_[k];
+			}
+			contact_density_[index_i] = contact_density_i;
+		}
+    //=========================================================================================================//
+    }
+}
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/for_2D_build/particle_dynamics/solid_dynamics/contact_dynamics_supplementary.cpp:40]: Why is this file new

File: SPHINXsys/src/for_3D_build/adaptations/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${PROJECT_SOURCE_DIR}/cmake) # main (top) cmake dir
+include(Headersearch)
+
+file(GLOB BASE_DIR_HEADERS *.h)
+DIR_INC_HEADER_NAMES("${BASE_DIR_HEADERS}" BASE_DIR_HEADER_NAMES)
+#message("${BASE_DIR_HEADER_NAMES}")
+INSTALL(FILES ${BASE_DIR_HEADER_NAMES} DESTINATION 3d_code/include)
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/for_3D_build/adaptations/CMakeLists.txt:7]: Why is this file new

File: SPHINXsys/src/for_3D_build/adaptations/adaptation_supplementary.cpp [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+					Vec3d particle_location(Real(i) * particle_spacing,
+											Real(j) * particle_spacing, Real(k) * particle_spacing);
+					Real distance = particle_location.norm();
+					if (distance < cutoff_radius)
+						sigma += kernel_ptr_->W(distance, particle_location);
+				}
+		return sigma;
+	}
+    //=================================================================================================//
+}
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/for_3D_build/adaptations/adaptation_supplementary.cpp:25]: Why is this file new

File: SPHINXsys/src/for_2D_build/adaptations/adaptation_supplementary.cpp (hunk 3 above)
Comment (ChiZhangatTUM) [reply SPHINXsys/src/for_2D_build/adaptations/adaptation_supplementary.cpp:23]: There are taken from shared/adaptations for consistency.

File: SPHINXsys/src/for_3D_build/materials/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,7 @@
+set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${PROJECT_SOURCE_DIR}/cmake) # main (top) cmake dir
+include(Headersearch)
+
+file(GLOB BASE_DIR_HEADERS *.h)
+DIR_INC_HEADER_NAMES("${BASE_DIR_HEADERS}" BASE_DIR_HEADER_NAMES)
+#message("${BASE_DIR_HEADER_NAMES}")
+INSTALL(FILES ${BASE_DIR_HEADER_NAMES} DESTINATION 3d_code/include)
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/for_3D_build/materials/CMakeLists.txt:7]: Why is this file new

File: SPHINXsys/src/for_3D_build/materials/elastic_solid_supplementary.cpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+		Lambda_(0,1) = Compliance_inv(0,1);
+		Lambda_(0,2) = Compliance_inv(0,2);
+		Lambda_(1,2) = Compliance_inv(1,2);
+		// the matrix is symmetric
+		Lambda_(1,0) = Lambda_(0,1);
+		Lambda_(2,0) = Lambda_(0,2);
+		Lambda_(2,1) = Lambda_(1,2);
+	}
+    //=================================================================================================//
+}
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/for_3D_build/materials/elastic_solid_supplementary.cpp:88]: Why is this file new

File: SPHINXsys/src/for_3D_build/particle_dynamics/solid_dynamics/contact_dynamics_supplementary.cpp [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+							}
+					}
+				}
+				contact_density_i += 0.005 * sigma * calibration_factor_[k];
+			}
+			contact_density_[index_i] = contact_density_i;
+		}
+        //=========================================================================================================//
+    }
+}
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/for_3D_build/particle_dynamics/solid_dynamics/contact_dynamics_supplementary.cpp:45]: Why is this file new

File: SPHINXsys/src/shared/bodies/probe_body.h [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+{
+	class ProbeBody : public SPHBody
+	{
+	public:
+		ProbeBody(SPHSystem &sph_system, SharedPtr<Shape> shape_ptr);
+		ProbeBody(SPHSystem &sph_system, const std::string &name);
+		virtual ~ProbeBody(){};
+	};
+}
+#endif // PROBE_BODY_H
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/shared/bodies/probe_body.h:51]: Why is this file new
Change of class name from observer -> probe

File: SPHINXsys/src/shared/particle_neighborhood/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,8 @@
+set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${PROJECT_SOURCE_DIR}/cmake) # main (top) cmake dir
+include(Headersearch)
+
+file(GLOB BASE_DIR_HEADERS *.h)
+DIR_INC_HEADER_NAMES("${BASE_DIR_HEADERS}" BASE_DIR_HEADER_NAMES)
+#message("${BASE_DIR_HEADER_NAMES}")
+INSTALL(FILES ${BASE_DIR_HEADER_NAMES} DESTINATION 2d_code/include)
+INSTALL(FILES ${BASE_DIR_HEADER_NAMES} DESTINATION 3d_code/include)
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/shared/particle_neighborhood/CMakeLists.txt:8]: Why is this file new

File: tests/2d_examples/test_2d_taylor_green_euler/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        set_target_properties(${PROJECT_NAME} PROPERTIES VS_DEBUGGER_WORKING_DIRECTORY "${EXECUTABLE_OUTPUT_PATH}")
+        target_link_libraries(${PROJECT_NAME} sphinxsys_2d ${TBB_LIBRARYS} ${Simbody_LIBRARIES})
+        add_dependencies(${PROJECT_NAME} sphinxsys_2d)
+else(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
+        if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
+                target_link_libraries(${PROJECT_NAME} sphinxsys_2d ${TBB_LIBRARYS} ${Simbody_LIBRARIES} ${Boost_LIBRARIES} stdc++)
+        else(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
+                target_link_libraries(${PROJECT_NAME} sphinxsys_2d ${TBB_LIBRARYS} ${Simbody_LIBRARIES} ${Boost_LIBRARIES} stdc++ stdc++fs)
+        endif(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
+endif(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
```
Comment (FabienPean-Virtonomy) [inline tests/2d_examples/test_2d_taylor_green_euler/CMakeLists.txt:40]: Why is this file new

**Review by FabienPean-Virtonomy — COMMENTED** [review]
There seems to be some weird things that happened during the merge. Could you please rebase your eigen branch (before doing the merge commit) onto the master ?

File: SPHINXsys/src/for_2D_build/particle_dynamics/solid_dynamics/contact_dynamics_supplementary.cpp (hunk 4 above)
Comment (ChiZhangatTUM) [reply SPHINXsys/src/for_2D_build/particle_dynamics/solid_dynamics/contact_dynamics_supplementary.cpp:40]: this function belong to 3D implementation.

File: SPHINXsys/src/for_3D_build/adaptations/adaptation_supplementary.cpp (hunk 6 above)
Comment (ChiZhangatTUM) [reply SPHINXsys/src/for_3D_build/adaptations/adaptation_supplementary.cpp:25]: This function belong to 3D  implemeantation.

File: SPHINXsys/src/for_3D_build/materials/elastic_solid_supplementary.cpp (hunk 8 above)
Comment (ChiZhangatTUM) [reply SPHINXsys/src/for_3D_build/materials/elastic_solid_supplementary.cpp:88]: They are generally related to 3D implementation.

File: SPHINXsys/src/for_3D_build/particle_dynamics/solid_dynamics/contact_dynamics_supplementary.cpp (hunk 9 above)
Comment (ChiZhangatTUM) [reply SPHINXsys/src/for_3D_build/particle_dynamics/solid_dynamics/contact_dynamics_supplementary.cpp:45]: They are generally related to 3D implementation. Otherwise ambiguous error is reported.

File: SPHINXsys/src/shared/particle_neighborhood/CMakeLists.txt (hunk 11 above)
Comment (ChiZhangatTUM) [reply SPHINXsys/src/shared/particle_neighborhood/CMakeLists.txt:8]: This is due to error of merging, should be deleted.

File: tests/2d_examples/test_2d_taylor_green_euler/CMakeLists.txt (hunk 12 above)
Comment (ChiZhangatTUM) [reply tests/2d_examples/test_2d_taylor_green_euler/CMakeLists.txt:40]: This is due to error of merging, should be deleted.

Comment (BenceVirtonomy) [reply tests/2d_examples/test_2d_taylor_green_euler/CMakeLists.txt:40]: This is why we suggest starting a clean PR - it's error-prone at the moment, and very hard to review

Comment (ChiZhangatTUM) [reply tests/2d_examples/test_2d_taylor_green_euler/CMakeLists.txt:40]: > This is why we suggest starting a clean PR - it's error-prone at the moment, and very hard to review

I will close this PR, and request a new one.
